File: typescript/adapter/weather-service-adaptee.ts
import { LegacyWeatherService } from "./legacy-weather-service";
import { TemperatureDate } from "./temperature-date";
import { WeatherServiceAdapter } from "./weather-service-adapter";

class WeatherServiceAdaptee implements WeatherServiceAdapter {
  private readonly legacyWeatherService: LegacyWeatherService;

  constructor(legacyWeatherService: LegacyWeatherService) {
    this.legacyWeatherService = legacyWeatherService;
  }

  getTemperature(longitude: number, latitude: number): TemperatureDate {
    const temperatureDataInXML = this.legacyWeatherService.getTemperature(
      this.getCityOf(longitude, latitude),
      this.getCountryOf(longitude, latitude)
    );
    return this.convertXMLDataToJson(temperatureDataInXML);
  }

  private convertXMLDataToJson(xmlData: string): TemperatureDate {
    console.log("Converting...");
    return new TemperatureDate("Converted Data from XML into JSON");
  }

  private getCityOf(longitude: number, latitude: number): string {
    console.log(
      `Extracting city of longitude: ${longitude} and latitude: ${latitude}`
    );
    return "City";
  }

  private getCountryOf(longitude: number, latitude: number): string {
    console.log(
      `Extracting country of longitude: ${longitude} and latitude: ${latitude}`
    );
    return "Country";
  }
}

File: Python/adaptor/WeatherServiceAdaptee.py
from WeatherServiceAdapter import WeatherServiceAdapter
from LegacyWeatherService import LegacyWeatherService
from TemperatureData import TemperatureData

class WeatherServiceAdaptee(WeatherServiceAdapter):
    def __init__(self, legacy_weather_service: LegacyWeatherService):
        self.legacy_weather_service = legacy_weather_service

    def get_temperature(self, longitude: float, latitude: float) -> TemperatureData:
        city = self.get_city_of(longitude, latitude)
        country = self.get_country_of(longitude, latitude)
        xml_data = self.legacy_weather_service.get_temperature(city, country)
        return self.convert_xml_to_json(xml_data)

    def convert_xml_to_json(self, xml_data: str) -> TemperatureData:
        print("Converting XML to JSON...")
        return TemperatureData("Converted Data from XML into JSON")

    def get_city_of(self, longitude: float, latitude: float) -> str:
        print(f"Extracting city for longitude: {longitude} and latitude: {latitude}")
        return "City"

    def get_country_of(self, longitude: float, latitude: float) -> str:
        print(f"Extracting country for longitude: {longitude} and latitude: {latitude}")
        return "Country"